Customize how long ships linger in a system

UnorderedSigh opened the pull request:
**Feature:** Allows customizing how long an npc will linger, or how long random interval fleets will linger.

## Feature Details

The `lingering` personality is a new one I added recently, which causes a ship to stay in a system for a while if it has nothing to do. Right now, ships will linger 1/4 of the minimum fleet arrival period of the system in which they were spawned (the number after `fleet` in the system).

For fleets with the `lingering` personality, you can now configure how long they linger.

```ruby
system Somewhere
    ...
    fleet 12000 "Large Syndicate"
    fleet 7000 "Small Core Pirate"
    fleet 30000 "Korath Raid"
    # Ships with the "lingering" personality will linger 60 seconds:
    "linger time" 3600

mission "My Mission"
    ...
    on offer
        dialog "Scan the <npc>. It'll stay in the system for 30 seconds."
    npc scan
        government "Merchant"
        ship "Korsmanath A'awoj" "Mousy"
        personality
            confusion 40
            # Ship needs the "lingering" personality for it to linger.
            uninterested timid frugal appeasing lingering
            # Ship will linger in the system 30 seconds before leaving
            "linger time" 1800
```

Obviously, configuring on a per-`fleet` basis in the `system` definition would be nice, but I would need #8082 first.

## Usage Examples
- Systems where raiding fleets lurk in wait for careless merchants.
- "Scan the NPC" missions give the player X seconds to react.

## Testing Done
1. Copy this to [lingertimetests.txt](https://github.com/endless-sky/endless-sky/files/11969681/lingertimetests.txt)
2. Visit the spaceport at Sunracer for instructions
3. Complete steps 1-5. You'll need scanners for the fifth test.

### Automated Tests Added
Can't think of any way to add one.

## Performance Impact
N/A

quyykk commented:
> Obviously, configuring on a per-fleet basis in the system definition would be nice,

Yes that's definitely better. I don't think it makes much sense to have this be a property of a system, so I would lean towards no, imo.

UnorderedSigh commented:
> Yes that's definitely better. I don't think it makes much sense to have this be a property of a system, so I would lean towards no, imo.

Lingering has to be system-dependent, or a system with high arrival rates will build up an infinite supply of ships. I tried the other way, and the systems are overwhelmed with ships.

quyykk commented:
So to be clear, this

> ships will linger 1/4 of the minimum fleet arrival period of the system

is too high for certain systems?

UnorderedSigh commented:
> is too high for certain systems?

Setting it to 1/4 led to a reasonable system activity rate, if I'm understanding @warp-core's test report. I think we could bump it up to 1/2 or even full, but that isn't what I'm tweaking in this PR.

The purpose of this PR is to give content creators the power to configure the system to be appropriate to the story. I'm implementing two of the three ideal configuration methods.

UnorderedSigh commented:
I'll explain this another way. If you have ships linger 2000 frames in a system with an arrival rate of 300, then there will be mountains of ships. If they linger 2000 frames in a system with an arrival rate of 30000, then there's no point in lingering at all.

If you want the lingering to both be useful and look right, you need it to be system-dependent.

One quarter of *the system's fleet arrival rate* is *system-dependent,* and by linking the lingering time to the system activity level. It leads to the appearance of ships waiting a reasonable amount of time. If you stay in a system for a while, you'll see they often successfully ambush ships due to the lingering.

So it does work as intended, but it isn't configurable enough for all purposes. You might want the Korath raiders, in their big, well-stocked ships, to be willing to wait twice as long as a small pirate ship. Or, for story reasons, pirates linger for a long time in this one system, but not so long in others.

quyykk commented:
Having it be configurable per fleet makes sense to me yeah. But making it configurable by system doesn't really, and I think I just don't get your explanations. 

> If you want the lingering to both be useful and look right, you need it to be system-dependent.

sure, just like it is now. But why make it configurable? Sorry if I'm misunderstanding something

1010todd commented:
Data organization wise it might NPC actions/personality might not belong in systems but in this case it should be in the system because it's directly related to how much NPC player will see in that system. We need to be able to configure at system level because if you set a fleet to linger for a minute but in one system they spawn every 10 seconds you're going to get a massive pile of ships lingering in that system. If "lingering time" can only be configured at fleet level we'll need new fleet for every system we want lingering time to be different which is far less ideal.

UnorderedSigh commented:
> But making it configurable by system doesn't really, and I think I just don't get your explanations.

I'm not sure what you're suggesting. The arrival rate must vary from system to system. A configuration method that doesn't allow that will rarely, if ever, be useful.

#### Useless feature: configure linger time in fleet definition

This would be a nearly useless feature. The linger time would be too long for "System 1" and too short for "System 2".

```ruby
fleet "Korath Raid"
  personality
    # There is no way to make this system-dependent.
    "linger time" 3000

system "System 1"
  ...
  fleet "Korath Raid" 30000
  fleet "Small Core Merchant" 30000
  # Korath raid would not linger long enough to wait for the 30000 period targets

system "System 2"
  ...
  fleet "Korath Raid" 10000
  fleet "Small Core Pirate" 500
  fleet "Large Core Pirate" 600
  # There would be a huge build-up of raiders. During the rare times when they have nothing to do,
  # they'd remain for the few frames it takes to have new prey.
```

#### Useful features

This, however, is useful:

```ruby
system Wherever
  ...
  # Suppose the first fleet doesn't linger, but the next two do linger.
  fleet "Something that never interacts with anyone, and never lingers" 4000
  fleet "Side 1, who hates side 2, and will linger" 30000
  fleet "Side 2, who hates side 1, and will linger" 30000
  # This would be the default:
  # "linger time" 1000
  # Content creator overrode it:
  "linger time" 12000
```
  
The default linger time would be 1000, because the default is based on the first fleet (4000/4=1000). It should be 7500 (30000=7500) because only the second and third fleets stay in the system to interact. The algorithm doesn't know that, so someone needs to override the calculation. The content creator decided to up the number to 12000 because they felt that longer linger time fit the lore of the system better.

This is also useful:

```ruby
mission "Do something"
   npc scan
      personality
         "linger time" 1000
```

It lets you control how long a ship lingers, based on the needs of the mission, rather than the target system activity level.

Specifying per fleet, per system, is more powerful, but it ignores one important type of ship: NPC ships. Any NPC ships that spawn in a system have the system's linger time. Without the ability to specify the system linger time, NPC ships will get 1/4 of the minimum arrival period.

```ruby
system Wherever
  ...
  fleet "Large Core Pirate" 5000
      "linger time" 1000
  fleet "Large Core Pirate" 3000
      "linger time" 1000
  fleet "Korath Raid" 30000
      # Korath linger longer than pirates
      "linger time" 8000
  fleet "Large Core Merchant" 4000
  fleet "Small Core Merchant" 2000
```

Notice that two of those linger times are the same. This is a more compact representation:

```ruby
system Wherever
  ...
  "linger time" 1000
  fleet "Korath Raid" 30000
      "linger time" 8000
  fleet "Large Core Pirate" 5000
  fleet "Large Core Pirate" 3000
  fleet "Large Core Merchant" 4000
  fleet "Small Core Merchant" 2000
```

Notice one other thing that does: any NPCs that spawn will *also* have the system linger time. Without `"linger time" 1000`, the NPCs would have a linger time of 500.

UnorderedSigh commented:
>  Also maybe the "linger time" could be child of the fleet in the system (indented and under stuffs like fleet "Korath Raid" 15000)? So different fleet could linger for different amount of time per system as well. Not just a system-wide linger time for everyone.

I hope my last post explained, in detail, why you need both. It covers all possible methods of configuration, and when each is useful or not. Plus, the one critical type of fleet you're forgetting: NPC fleets that are spawned in the system. They have the system linger time. If you have no way to override that, NPCs will spawn with 1/4 of the minimum arrival time.

UnorderedSigh commented:
> Also maybe the "linger time" could be child of the fleet in the system (indented and under stuffs like fleet "Korath Raid" 15000)? So different fleet could linger for different amount of time per system as well. Not just a system-wide linger time for everyone.

One other reason to have the system-level "linger time" is so you don't have to know which fleets linger. Sure, you might know which ones linger on the day you PR your changes. What if someone adds `lingering` to a fleet? Should they change the 100 systems where that fleet appears, to hand-tune the linger time? What if an `event` adds `lingering` to a fleet? Should it also change the 100 systems where the fleet appears?

quyykk commented:
> but it ignores one important type of ship: NPC ships

For this you could add a node for NPCs that would override the linger time, similar to how it's done for cargo. Would that resolve your issue?

UnorderedSigh commented:
> For this you could add a node for NPCs that would override the linger time, similar to how it's done for cargo. Would that resolve your issue?

NPCs need a way to have system-dependent linger times, too. That can only happen if the tag is in the `system` declaration.

UnorderedSigh commented:
> For this you could add a node for NPCs that would override the linger time, similar to how it's done for cargo. Would that resolve your issue?

It also would not solve this problem:

> One other reason to have the system-level "linger time" is so you don't have to know which fleets linger. Sure, you might know which ones linger on the day you PR your changes. What if someone adds `lingering` to a fleet? Should they change the 100 systems where that fleet appears, to hand-tune the linger time? What if an `event` adds `lingering` to a fleet? Should it also change the 100 systems where the fleet appears?

UnorderedSigh commented:
Last tested and known to work Mon, 29 May 2023 17:29:10 UTC

Hurleveur commented on source/System.cpp:
```diff
@@ -966,6 +974,13 @@ int System::MinimumFleetPeriod() const
 
+int System::LingerTime() const
+{
+	return lingerTime >= 0 ? lingerTime : minimumFleetPeriod;
```
It used to be minimumfleetperiod/4, why the change?

UnorderedSigh replied:
That's an error. It should be `minimumFleetPeriod / 4`.

UnorderedSigh replied:
And now it is.

UnorderedSigh commented:
Last tested and known to work Thu, 06 Jul 2023 15:45:53 UTC.

UnorderedSigh commented:
Last tested and known to work Sun, 27 Aug 2023 15:37:00 UTC

Hurleveur commented on source/System.cpp:
```diff
@@ -209,6 +211,12 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 			child.PrintTrace("Error: Expected key to have a value:");
 			continue;
 		}
+		else if(remove && child.Size() == 2 && key == "linger time")
+			lingerTime = -1;
```
isnt this already taken care of right above?

UnorderedSigh replied:
Maybe? I'll try removing it and see what happens.

UnorderedSigh replied:
It is gone now.

Hurleveur commented on source/System.cpp:
```diff
@@ -209,6 +211,12 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 		}
+		else if(remove && child.Size() == 2 && key == "linger time")
+			lingerTime = -1;
+		else if(!remove && child.Size() == valueIndex + 1 && key == "linger time")
+			lingerTime = child.Value(valueIndex);
+		else if(key == "linger time")
+			child.PrintTrace("Invalid \"linger time\" specification.");
```
prob should have both of these in the same else if(key == "linger time") for it to be a bit clearer

UnorderedSigh replied:
That would add two lines, and another nesting level. People have preferred an else-if chain in the past.

UnorderedSigh replied:
I could remove the "invalid linger time specification" check entirely, and it'll fall through to "skipping unrecognized attribute." Not as informative, but it does shrink the code.

Hurleveur replied:
Your call, this isnt anything major

UnorderedSigh replied:
It's gone now.

Hurleveur commented on source/System.cpp:
```diff
@@ -966,6 +974,13 @@ int System::MinimumFleetPeriod() const
 
+int System::LingerTime() const
+{
+	return lingerTime >= 0 ? lingerTime : minimumFleetPeriod / 4;
```
we could also initialize it at the end of load so we dont have to calc this each time?

UnorderedSigh replied:
No, if it is to be cached, it belongs in `UpdateSystem()`

```C++
// Update any information about the system that may have changed due to events,                                                                                                                                                                                                   
// or because the game was started, e.g. neighbors, solar wind and power, or                                                                                                                                                                                                      
// if the system is inhabited.                                                                                                                                                                                                                                                    
void System::UpdateSystem(const Set<System> &systems, const set<double> &neighborDistances)
```

UnorderedSigh replied:
Okay, now I have:
- `requestedLingerTime` = the data file "linger time"
- `actualLingerTime` = a cached calculation in UpdateSystem()
The AI uses the actualLingerTime, of course.

Hurleveur replied:
I mean since it was 4 you were dividing by it was prob fine-ish but good, every opti counts

Hurleveur reviewed (changes requested):
Looks good and pretty simple changed making it more useful, I just noticed a few things that I think could be improved upon

UnorderedSigh commented:
The Windows Clang tests are failing for a nonsensical reason that appears to have nothing to do with my changes.

```
   FAILED: source/CMakeFiles/EndlessSkyLib.dir/Account.cpp.obj 
  sccache C:\PROGRA~1\LLVM\bin\clang-cl.exe  /nologo -TP -DUNICODE -D_UNICODE -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include\SDL2 -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include\AL /DWIN32 /D_WINDOWS /EHsc /O2 /Ob2 /DNDEBUG -MD -flto=thin /W3 /permissive- /sdl /GR- /analyze- -Wno-nonportable-include-path /Gy /Oi -WX /showIncludes /Fosource\CMakeFiles\EndlessSkyLib.dir\Account.cpp.obj /Fdsource\CMakeFiles\EndlessSkyLib.dir\ -c -- D:\a\endless-sky\endless-sky\source\Account.cpp
  clang-cl: error: no such file or directory: '/wd4668'
```

Hurleveur commented on source/System.cpp:
```diff
@@ -213,6 +215,10 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
+		else if(!remove && child.Size() == valueIndex + 1 && key == "linger time")
+			requestedLingerTime = child.Value(valueIndex);
+		else if(key == "linger time")
+			child.PrintTrace("Invalid \"linger time\" specification.");
```
Now that you mention it we prob dont need this since the one at the end will pretty much say the same thing

UnorderedSigh replied:
That would only save two lines, and it would make the message less informative.

Hurleveur replied:
then I think you should give the proper specification or explain what is invalid
because honestly there isnt much difference between `Skipping unrecognized attribute: "linger time"` and this

UnorderedSigh replied:
The full information would be too long. I removed it instead. Should someone ask me to put it back, I'll come up with something.

Hurleveur reviewed (commented):
I hadnt realized there was a linger time for systems and npcs available, sounds useful

UnorderedSigh commented:
Got the same Windows Clang error:

```
  sccache C:\PROGRA~1\LLVM\bin\clang-cl.exe  /nologo -TP  -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include\SDL2 -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include\AL /DWIN32 /D_WINDOWS /EHsc /O2 /Ob2 /DNDEBUG -MD -flto=thin -WX /showIncludes /FoCMakeFiles\EndlessSky.dir\source\main.cpp.obj /FdCMakeFiles\EndlessSky.dir\ -c -- D:\a\endless-sky\endless-sky\source\main.cpp
  clang-cl: error: no such file or directory: '/wd4668'
  ninja: build stopped: subcommand failed.
```

UnorderedSigh commented:
And a third time, the same error. This is probably not a coincidence.

```
   [4/198] Building CXX object source\CMakeFiles\EndlessSkyLib.dir\Account.cpp.obj
  FAILED: source/CMakeFiles/EndlessSkyLib.dir/Account.cpp.obj 
  sccache C:\PROGRA~1\LLVM\bin\clang-cl.exe  /nologo -TP -DUNICODE -D_UNICODE -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include\SDL2 -imsvcD:\a\endless-sky\endless-sky\build\ci\vcpkg_installed\x64-windows-release\include\AL /DWIN32 /D_WINDOWS /EHsc /O2 /Ob2 /DNDEBUG -MD -flto=thin /W3 /permissive- /sdl /GR- /analyze- -Wno-nonportable-include-path /Gy /Oi -WX /showIncludes /Fosource\CMakeFiles\EndlessSkyLib.dir\Account.cpp.obj /Fdsource\CMakeFiles\EndlessSkyLib.dir\ -c -- D:\a\endless-sky\endless-sky\source\Account.cpp
  clang-cl: error: no such file or directory: '/wd4668'
```

UnorderedSigh commented:
Apparently there are some known issues with the current Azure windows images. It'll be fixed in a week or so. This is completely outside of our control. We'll just have to wait.

Hurleveur reviewed (approved):
Good addition, and looks ok to me code wise
I havent tested it but you did so all good on that side too

UnorderedSigh commented:
Last tested and known to work Wed, 22 Nov 2023 23:26:14 UTC

warp-core commented on source/Personality.cpp:
```diff
@@ -130,27 +130,32 @@ void Personality::Load(const DataNode &node)
 			else
 				confusionMultiplier = child.Value(1);
 		}
+		else if(remove && child.Size() == 2 && child.Token(1) == "linger time")
+			lingerTime = -1;
+		else if(!remove && child.Size() == 2 + keyIndex && child.Token(keyIndex) == "linger time")
+			lingerTime = child.Value(add + 1);
+		else if(child.Size() > keyIndex && child.Token(keyIndex) == "linger time")
+			child.PrintTrace("Invalid \"linger time\" specification.");
```
```c++
		else if(remove && child.Size() == 2 && child.Token(1) == "linger time")
			lingerTime = -1;
		else if(!remove && child.Size() == 2 + keyIndex && child.Token(keyIndex) == "linger time")
			lingerTime = child.Value(add + 1);
		else if(child.Size() > keyIndex && child.Token(keyIndex) == "linger time")
			child.PrintTrace("Invalid \"linger time\" specification.");
```

I'm not really sure what's going on here.
We're currently looking at the children of the `personality` node.
The parent `personality` node may take any of the following forms:
```
personality [flags...]
add personality [flags...]
remove personality [flags...]
```
The `flags` are any of the personalities. `confusion` is not allowed there, and neither is `"linger time"`. Not important here.
The bool `add` will be true if the first token in that parent is `add`, false otherwise. The bool `remove` will be true if that first token is `remove`, false otherwise.
Looking at the child nodes, we could have:
```
personality
	"linger time" <time#> # assigns the time to the linger time variable.
add personality
	"linger time" <time#> # assigns the time to the linger time variable.
remove personality
	"linger time" # sets the linger time variable to -1. Removing the linger time so no value needs to be provided.
```
The first branch of your code seems to deal with the `remove` case.
```c++
		else if(remove && child.Size() == 2 && child.Token(1) == "linger time")
			lingerTime = -1;
```
If checks if `remove` is true, ok. But then, it seems to expect the child to have two tokens, and the second of those to be `"linger time"`. This seems to expect the syntax:
```
	remove "linger time"
```
But, the `remove` bool you're checking was set based on the presence of `remove` on the `personality` node, not in this child.
If we want:
```
personality
	remove "linger time"
```
Instead of:
```
remove personality
	"linger time"
```
Then this code needs to be updated to check the first token of the child for `remove`, not reuse the check on the parent node.
Having `remove "linger time"` when we don't have `remove <flag>` seems like it'll be confusing, so I think we should stick to `remove` going on the parent node.
In that case, this first branch should be:
```c++
		else if(remove && child.Token(0) == "linger time")
			lingerTime = -1;
```
I don't think we need to consider this whole child node invalid if someone's done:
```
remove personality
	"linger time" 500
```
Or similar, but a warning about the additional token not having any effect might be useful:
```c++
		else if(remove && child.Token(0) == "linger time")
		{
			lingerTime = -1;
			if(child.Size() > 1)
				// Print a warning or something about the additional token(s) being ignored.
		}
```

```c++
		else if(!remove && child.Size() == 2 + keyIndex && child.Token(keyIndex) == "linger time")
			lingerTime = child.Value(add + 1);
```
The next branch assigns the given value to the `lingerTime` variable. But it's using the index variable that is based onthe parent node, not this node.
Say we had:
```
add personality
	"linger time" 500
```
`keyIndex` is going to be `1`, and `add` is going to be true.
So, this is expecting the child to have three tokens, and for `"linger time"` to be the second, and the value to be the third. Even if someone did:
```
personality
	add "linger time" 500
```
That still wouldn't work because `keyIndex` is now `0` and `add` is false, so it's expecting two tokens in the child node, with `"linger time"` being the first and the value being the second.
This should instead be something like:
```c++
		else if(!remove && child.Size() == 2 && child.Token(0) == "linger time")
			lingerTime = child.Value(1);
```

```c++
		else if(child.Size() > keyIndex && child.Token(keyIndex) == "linger time")
			child.PrintTrace("Invalid \"linger time\" specification.");
```
We generally don't discard something entirely just because it has too many tokoens. This is also looking at the wrong index for `"linger time"`.

I think the conditions should also be reordered and placed within a nested branch.
```suggestion
		else if(child.Token(0) == "linger time)
		{
			if(remove)
				lingerTime = -1;
			else if(child.Size() >= 2)
				lingerTime = child.Value(1);
			else
				child.PrintTrace("Invalid \"linger time\" specification.");
		}
```

UnorderedSigh replied:
The linger time isn't always at index 0, so that wouldn't work. I see what you're getting at, and I'll try to simplify this code.

UnorderedSigh replied:
I've applied your suggested change manually, except: 1) I added the missing `"` and 2) The error message is more specific.

warp-core commented on source/System.h:
```diff
@@ -174,6 +174,13 @@ class System {
 	// The smallest arrival period of a fleet (or 0 if no fleets arrive)
 	int MinimumFleetPeriod() const;
 
+	// The value of "linger time" from the data files:
+	int RequestedLingerTime() const;
```
Does this method need to exist? It doesn't appear to be used.

warp-core commented on source/System.h:
```diff
@@ -174,6 +174,13 @@ class System {
 	// The smallest arrival period of a fleet (or 0 if no fleets arrive)
 	int MinimumFleetPeriod() const;
```
This method also seems to not be used anymore.

warp-core commented on source/System.cpp:
```diff
@@ -213,6 +215,8 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 			child.PrintTrace("Error: Expected key to have a value:");
 			continue;
 		}
+		else if(!remove && child.Size() == valueIndex + 1 && key == "linger time")
+			requestedLingerTime = child.Value(valueIndex);
```
The case of `remove "linger time"` should have been handled by this point.
Similarly, it's already ensured that this node has a value. Just skipping it entirely if there are extraneous tokens after that isn't something we normally do.
```suggestion
		else if(key == "linger time")
			requestedLingerTime = child.Value(valueIndex);
```

UnorderedSigh replied:
The case of `remove (anything)` isn't handled up to that point. That giant "else if" chain handles adding and removing. Only the case of removing everything has been handled.

Checking the child size handles the special case of `add "linger time"` without a value.

I made it look nicer. How do you like it now?

warp-core commented on source/System.cpp:
```diff
@@ -526,6 +530,8 @@ void System::UpdateSystem(const Set<System> &systems, const set<double> &neighbo
 		minimumFleetPeriod = min<int>(minimumFleetPeriod, event.Period());
 	if(minimumFleetPeriod == numeric_limits<int>::max())
 		minimumFleetPeriod = 0;
+
+	actualLingerTime = requestedLingerTime >= 0 ? requestedLingerTime : minimumFleetPeriod / 4;
```
Does `requestedLingerTime` even need to exist at all? It appears to only ever be used in this method and the unused getter method.
It doesn't seem to need to be a class member. The other method that refers to it is never used and so doesn't need to exist.
This method does use it, but it doesn't seem to need to.
Why not just have a single `lingerTime` variable with the default value being `-1`.
Then, if no value is given for it, it will still be negative and you can do this same thing just with only that variable.
```suggestion

	if(lingerTime < 0)
	lingerTime = minimumFleetPeriod / 4;
```

UnorderedSigh replied:
That wouldn't work. If the minimum fleet arrival period changed, the lingerTime wouldn't be updated since it's not -1 anymore.

Consider this:

1. A system exists.
2. Add "linger time" to the system.
3. Linger time is now 1/4 of the minimum fleet arrival period.
4. Change the fleets in the system.
5. Linger time isn't 1/4 of the minimum fleet arrival period anymore because System forgot the -1 value.

The separate actualLingerTime and requestedLingerTime allows the System to remember it should use the 1/4 arrival period value, and automatically update that when the minimum arrival period changes.

TomGoodIdea commented:
@UnorderedSigh Do you want to continue working on this, or should we mark it as "orphaned"?

UnorderedSigh commented:
> Do you want to continue working on this, or should we mark it as "orphaned"?

I'll get to this one soon.

UnorderedSigh commented:
I've synced this with master, but I haven't addressed the comments yet. It sounds like there's some problems with how I implemented the `linger time` property.

UnorderedSigh commented:
I've addressed the reviewer comments and cleaned up the code.

Last tested and known to work Tue, 27 Aug 2024 16:59:52 UTC.

warp-core reviewed (commented):
Use an `std::optional" for the linger time values instead of a sentinel of -1.

TheGiraffe3 commented:
Closing as orphaned because this PR has been waiting on OP for several months.
